src/main.rs

```diff
@@ -0,0 +1,253 @@
+            }
+            _ => panic!("Unsupported message type"),
+        }
+    } else {
+        match cmd_type {
+            CommandType::FirmwareVersion => {
+                from_json_to_wire::<
+                    firmware_version::FirmwareVersionResponse,
+                    _,
+                    _,
+                >(input_reader, stdwrite)
+            }
+            _ => panic!("Unsupported message type"),
+        }
+    }
+}
+
+/// Macro to deserialize wire format from an input file and then run an operation on the
+/// deserialized message.
+/// Arguments:
+/// * `input_file`: Identifier to get the input file name from.
+/// * `msg`: Identifier to hold the message to operator.
+/// * `op`: Expression to execute after deserializing `msg` from `input_file`.
+macro_rules! read_wire_and_operate {
+    ($input_file:ident, $msg:ident, $op:expr) => {
+        let mut input = OpenOptions::new().read(true).open(&$input_file).expect("failed to open input file");
+
+        let mut read_buf = Vec::new();
+        input.read_to_end(&mut read_buf).expect("couldn't read from file");
+
+        let mut read_buf_slice = read_buf.as_slice();
+        let header =
+            Header::from_wire(&mut read_buf_slice).expect("failed to read header");
+        if header.is_request {
+          match header.command {
+            CommandType::FirmwareVersion => {
+              let $msg : protocol::firmware_version::FirmwareVersionRequest =
+                  FromWire::from_wire(&mut read_buf_slice).expect("failed to read request");
+
+              $op
+            },
+            _ => {
+              panic!("unsupported request type {:?}", header.command);
+            }
+          }
+        } else {
+          match header.command {
+            CommandType::FirmwareVersion => {
+              let $msg : protocol::firmware_version::FirmwareVersionResponse =
+                  FromWire::from_wire(&mut read_buf_slice).expect("failed to read response");
+
+              $op
+            },
+            _ => {
+              panic!("unsupported response type {:?}", header.command);
+            }
+          }
+        }
+    }
+}
+
+/// Deserializes a header + message in wire format from `input_file` and then serialize the
+/// message as JSON to `output_file`.
+fn to_json(input_file: &str, output_file: &str) {
+    let output = OpenOptions::new()
+        .write(true)
+        .truncate(true)
+        .create(true)
+        .open(&output_file)
+        .expect("failed to open output file");
+    read_wire_and_operate!(input_file, msg, {
+        serde_json::to_writer(output, &msg)
+            .expect("failed to serialize to JSON")
+    });
+}
+
+/// Deserializes a header + message in wire format from `input_file` and then prints the message
+/// to the console.
+fn print_message(input_file: &str) {
+    read_wire_and_operate!(input_file, msg, { eprintln!("{:?}", msg) });
+}
+
+fn main() {
+    let app = App::new("Manticore Tool")
```

Comment:
We can keep this like this for now, but I might come back later and clean this up to use `structopt` instead (which may or may not be a bit more readable than a huge builder...).

Reply:
ACK.